nexus/src/app/sagas/instance_update/mod.rs

```diff
@@ -0,0 +1,661 @@
+                "instance_update" => %"active VMM destroyed",
+            );
+        }
+        Err(err) => return Err(ActionError::action_failed(err)),
+    };
+
+    Ok(())
+}
+
+pub(super) async fn siu_unassign_oximeter_producer(
+    sagactx: NexusActionContext,
+) -> Result<(), ActionError> {
+    let osagactx = sagactx.user_data();
+    let RealParams { ref serialized_authn, ref authz_instance, .. } =
+        sagactx.saga_params::<RealParams>()?;
+
+    let opctx =
+        crate::context::op_context_for_saga_action(&sagactx, serialized_authn);
+
+    info!(
+        osagactx.log(),
+        "instance update (VMM destroyed): unassigning oximeter producer";
+        "instance_id" => %authz_instance.id(),
+        "instance_update" => %"active VMM destroyed",
+    );
+    crate::app::oximeter::unassign_producer(
+        osagactx.datastore(),
+        osagactx.log(),
+        &opctx,
+        &authz_instance.id(),
+    )
+    .await
+    .map_err(ActionError::action_failed)
+}
+
+async fn siu_update_and_unlock_instance(
+    sagactx: NexusActionContext,
+) -> Result<(), ActionError> {
+    let RealParams { serialized_authn, authz_instance, ref update, .. } =
+        sagactx.saga_params::<RealParams>()?;
+    unlock_instance_inner(
+        &serialized_authn,
+        &authz_instance,
+        &sagactx,
+        Some(&update.new_runtime),
+    )
+    .await?;
+
+    let opctx =
+        crate::context::op_context_for_saga_action(&sagactx, &serialized_authn);
+    let osagactx = sagactx.user_data();
+
+    // fetch the state from the database again to see if we should immediately
+    // run a new saga.
+    // TODO(eliza): go back and make the unlock-instance query return the
+    // current state, instead...
+    let new_state = match osagactx
+        .datastore()
+        .instance_fetch_all(&opctx, &authz_instance)
+        .await
+    {
+        Ok(s) => s,
+        Err(e) => {
+            warn!(osagactx.log(), "instance update: failed to fetch state on saga completion";
+                "instance_id" => %authz_instance.id(),
+                "error" => %e);
+            // if we can't refetch here, don't unwind all the work we did do.
+            // the instance-updater background task will take care of it.
+            return Ok(());
+        }
+    };
+
+    if UpdatesRequired::for_snapshot(osagactx.log(), &new_state).is_some() {
+        if let Err(e) = osagactx
+            .nexus()
+            .sagas
+            .saga_execute::<SagaInstanceUpdate>(Params {
+                // everyone in the friend group just venmo-ing the same
+                // serialized_authn back and forth forever.
+                serialized_authn,
+                authz_instance,
+            })
+            .await
+        {
+            // again, if this fails, don't unwind all the good work we already did.
+            warn!(osagactx.log(), "instant update: subsequent saga execution failed"; "error" => %e);
+        }
+    }
```

Comment:
This is, essentially, the jankiest possible implementation of having completed instance-update sagas chain into new sagas when more VMM/migration updates have been applied while the saga was running. I'd like to clean this up a bit --- in particular, this is currently split across two queries to release the lock and then re-fetch the state records, and I'd like to go back and change that.  But, I wanted to make sure that doing this actually resolved the raciness I was seeing in some of the tests, first, so for now we have the kind of bad version.